### PR TITLE
MAYA-131348 - Adapt to Maya Ufe world node change

### PR DESCRIPTION
**Pull request by seando-adsk**
MAYA-131348 - Nothing prevents users from naming nodes "world", which breaks USD workflow

* Incorrect to create Ufe::Path for Maya (hidden) world node using Ufe::PathString::path().
* Let DuplicateCommand accept empty destination path which means Maya (hidden) world node.
* Let CreateStageWithNewLayer command accept null input parent which means Maya world node.
* Use ufe.PathString.string() instead of str()

### CODE REVIEW

File: lib/mayaUsd/commands/PullPushCommands.cpp [hunk 1]
```diff
@@ -416,14 +426,13 @@ MStatus DuplicateCommand::doIt(const MArgList& argList)
...
-    if (status != MS::kSuccess)
-        return status;
+    CHECK_MSTATUS_AND_RETURN_IT(status);
 
     status = parseUfePathArg(argParser, 0, _srcPath);
     if (status != MS::kSuccess)
         return reportError(status);
 
-    status = parseUfePathArg(argParser, 1, _dstPath);
+    status = parseUfePathArg(argParser, 1, _dstPath, true /*allowEmpty*/);
```
Comment (seando-adsk) [inline lib/mayaUsd/commands/PullPushCommands.cpp:435]: I set new `allowEmpty` bool here to be able to accept as command input "" which means Maya "world" node.

File: lib/mayaUsd/commands/PullPushCommands.cpp [hunk 2]
```diff
@@ -455,12 +463,19 @@ MStatus DuplicateCommand::doIt(const MArgList& argList)
...
+                childPath = _dstPath.empty() ? ps : _dstPath + ps;
+            } else {
                 // USD duplicate to Maya
-                : _dstPath + _srcPath.back();
+                if (_dstPath.empty()) {
+                    childPath = Ufe::PathString::path("|" + _srcPath.back().string());
+                } else {
+                    childPath = _dstPath + _srcPath.back();
+                }
+            }
```
Comment (seando-adsk) [inline lib/mayaUsd/commands/PullPushCommands.cpp:478]: Had to rework logic to handle when _dstPath is empty.

File: lib/mayaUsd/fileio/primUpdaterManager.cpp [hunk 3]
```diff
@@ -554,10 +554,7 @@ using UsdPathToDagPathMapPtr = std::shared_ptr<UsdPathToDagPathMap>;
 using PushCustomizeSrc
     = std::tuple<SdfPath, UsdStageRefPtr, SdfLayerRefPtr, UsdPathToDagPathMapPtr>;
 
-PushCustomizeSrc pushExport(
-    const Ufe::Path&                 ufePulledPath,
-    const MObject&                   mayaObject,
-    const UsdMayaPrimUpdaterContext& context)
+PushCustomizeSrc pushExport(const MObject& mayaObject, const UsdMayaPrimUpdaterContext& context)
```
Comment (seando-adsk) [inline lib/mayaUsd/fileio/primUpdaterManager.cpp:557]: Removed unused input var `ufePullPath`

File: lib/mayaUsd/fileio/primUpdaterManager.cpp [hunk 4]
```diff
@@ -1560,8 +1557,10 @@ bool PrimUpdaterManager::duplicate(
     const Ufe::Path&    dstPath,
     const VtDictionary& userArgs)
 {
-    MayaUsdProxyShapeBase* srcProxyShape = MayaUsd::ufe::getProxyShape(srcPath);
-    MayaUsdProxyShapeBase* dstProxyShape = MayaUsd::ufe::getProxyShape(dstPath);
+    MayaUsdProxyShapeBase* srcProxyShape
+        = srcPath.empty() ? nullptr : MayaUsd::ufe::getProxyShape(srcPath);
+    MayaUsdProxyShapeBase* dstProxyShape
+        = dstPath.empty() ? nullptr : MayaUsd::ufe::getProxyShape(dstPath);
```
Comment (seando-adsk) [inline lib/mayaUsd/fileio/primUpdaterManager.cpp:1563]: Either of these is now allowed to be null.

File: lib/mayaUsd/ufe/MayaUsdContextOps.cpp [hunk 5]
```diff
@@ -683,9 +683,10 @@ Ufe::UndoableCommand::Ptr MayaUsdContextOps::doOpCmd(const ItemPath& itemPath)
     } else if (itemPath[0] == kEditAsMayaOptionsItem) {
         executeEditAsMayaOptions(path());
     } else if (itemPath[0] == kDuplicateAsMayaItem) {
+        // Note: empty string for target means Maya (hidden) world node.
         MString script;
         script.format(
-            "^1s \"^2s\" \"|world\"",
+            "^1s \"^2s\" \"\"",
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/MayaUsdContextOps.cpp:689]: Replace "world" with empty string which is now handled in command.

**Review by seando-adsk — COMMENTED** [review]
The code wasn't "incorrect" before its just that we were relying on knowledge of the Maya `hidden` world node (which is the root). Removed all that world knowledge except in very special cases. These changes are required now before I make a fix in Maya which will allow users to create their own top-level "world" node and use that with Ufe. Currently they cannot because of a bug in Maya.

**Review by pierrebai-adsk — APPROVED** [review]
Great!

Only downside is passing an empty string is not super-obvious when reading code, but nto much we can do about it.

**Comment by seando-adsk** [comment]
Yeah I added a few comments in those places to explain the empty string.

**Pull request merged**